qiskit/_result.py

```diff
@@ -253,4 +318,88 @@ def get_qubitpol_vs_xval(self, xvals_dict=None):
             for qubit_ind in range(nqubits):
                 qubitpol[circuit_ind,qubit_ind] = self.average_data(self.__qobj['circuits'][circuit_ind]['name'], z_dicts[qubit_ind])
 
-        return qubitpol,xvals
\ No newline at end of file
+        return qubitpol,xvals
+
+    def save_datestr(self, folder, fileroot):
+        """Constructs a filename using the current date-time
+
+        Args:
+            folder: path to the save folder
+            fileroot: root string for the file
+
+        Returns:
+            filename: full file path of the form 'folder/YYYY_MM_DD_HH_MM_fileroot.json'
+        """
+
+        #if the fileroot has .json appended strip it off
+        if (len(fileroot)>4 and fileroot[-5:].lower()=='.json'):
+            fileroot = fileroot[0:-5]
+
+        return os.path.join(folder,'{:%Y_%m_%d_%H_%M_}'.format(datetime.datetime.now())+fileroot+'.json')
+
+    def save(self, filename, metadata=None):
+        """Save a result (qobj + result) to a single dictionary file filename.json
+        If the file already exists then numbers will be appended to generate a unique filename
+
+        Args:
+            filename: save path
+            metadata (optional): Add another dictionary with custom data for the result (eg fit results)
+
+        Returns:
+            filename: full file path
+        """
+        master_dict = {}
+        master_dict['qobj'] = self.__qobj.copy()
+        master_dict['result'] = self.__result.copy()
+        if metadata is None:
+            master_dict['metadata'] = {}
+        else:
+            master_dict['metadata'] = metadata
+
+
+        #need to convert any ndarray variables to lists so that they can be
+        #exported to the json file
+        self._convert_qobj_to_json(master_dict['result'])
+
+        #if the filename has .json appended strip it off
+        if filename[-5:].lower()=='.json':
+            filename = filename[0:-5]
+
+        append_str = ''
+        append_num = 0
+
+        while (os.path.exists(filename+append_str+'.json')):
+            append_num += 1
+            append_str = '_%d'%append_num
+
+        fo = open(filename+append_str+'.json','w')
+        json.dump(master_dict,fo,indent=1)
+        fo.close()
+
+        return filename+append_str+'.json'
+
+    def load(self, filename):
```

Comment:
Related to the comment on the `Result.__init__()` changes, can we change this method to a class method that returns the newly constructed `Result`, renaming it a bit in the process? Something along the lines of:

```python
@classmethod
def from_json(cls, filename):
    ...
    return cls(result, qobj)
```

So the usage would be:
```python
my_loaded_result = Result.from_json('my_json_file.json')
```

The main drawback would be that `metadata` would not be returned, but the proposed change would make sure that all `Result` instantations would go through the constructor, which reduces the complexity by streamlining the instantiation into a single point which would be easy to change in the future and avoids having `Results` on an invalid state (ie. with `None` attributes). Even if that state would be only for a short while on its lifetime, doing it through the constructor avoids having to treat the loading as a separate case and reduces the chance of forgetting to update it in the future and similar implications.

For solving the issue of retrieving `metadata`, I'm unsure about the best option, but perhaps a reasonable option would be to have a `get_metadata_from_json()` method (ideally outside the class, as a function module, as it does not really use `Result`) that only parses the file and returns the converted `metadata` field.